### PR TITLE
Diffusion class

### PR DESCRIPTION
**Pull request by szjasinski**

### CODE REVIEW

File: utils.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    subset = Subset(dataset, selected_indices)
+
+    return subset
+
+
+def load_transformed_CIFAR10_automobiles(batch_size, max_samples=100, img_size=32):
+    data_transforms = [
+        transforms.Resize((img_size, img_size)),  # Resize to specified img_size
+        transforms.ToTensor(),  # Convert to tensor and scale to [0, 1]
+        transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5)),  # Normalize for CIFAR-10
```
Comment (Ulvi-Movs) [inline utils.py:29]: why we normalize to (0.5,0.5,0.5) not any other value?

File: utils.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    data_transform = transforms.Compose(data_transforms)
+    train_set = load_CIFAR10_automobiles(data_transform, max_samples=max_samples, train=True)
+    test_set = load_CIFAR10_automobiles(data_transform, train=False)
+    data = ConcatDataset([train_set, test_set])
+    dataloader = DataLoader(data, batch_size=batch_size, shuffle=True, drop_last=True)
+
+    return data, dataloader
+
+
+def visualize_images(images: list[torch.Tensor], n_rows: int, n_cols: int) -> None:
```
Comment (Ulvi-Movs) [inline utils.py:41]: if you make typing please make so on every functions, or not do at all =)

File: utils.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import torchvision
+import torchvision.transforms as transforms
+from torch.utils.data import DataLoader, ConcatDataset
+from torch.utils.data import DataLoader, Subset
+
+import matplotlib.pyplot as plt
+
+
+def load_CIFAR10_automobiles(data_transform, max_samples=100, train=True):
+    dataset = torchvision.datasets.CIFAR10(
```
Comment (Ulvi-Movs) [inline utils.py:11]: please add documentation

File: utils.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            img = images[idx][row].permute(1, 2, 0).cpu().numpy()
+            img = img * 0.5 + 0.5  # Denormalize
+            img = img.clip(0, 1)
+            
+            axes[row][col].imshow(img)
+            axes[row][col].axis("off")
+
+    plt.tight_layout()
+    plt.show()
+
```
Comment (Ulvi-Movs) [inline utils.py:68]: i think we can add here parametr: save_result - and if True save it in file. For feature i think it will be good idea to had such kind of possibilities.

File: diffusion.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        self.model = UNet(3, 3)
+
+
+    def _create_coeffs_from_betas(self, betas: torch.Tensor) -> dict[str, torch.Tensor]:
+        coeffs = dict()
+        alphas = 1. - betas
+        alphas_bar = torch.cumprod(alphas, dim=0)
+
+        # for forward process q(x_t | x_t-1)
+        coeffs["sqrt_one_minus_b"] = torch.sqrt(1 - betas)
```
Comment (Ulvi-Movs) [inline diffusion.py:35]: i think its not same with formula - we need correct formula or code? =)

File: diffusion.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        coeffs["sqrt_b"] = torch.sqrt(betas) # also for reverse process (substituted as sigmas)
+        
+        # for forward process q(x_t | x_0)
```
Comment (Ulvi-Movs) [inline diffusion.py:38]: same for here?

File: diffusion.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        coeffs["sqrt_a_bar"] = torch.sqrt(alphas_bar)
+        coeffs["sqrt_one_minus_a_bar"] = torch.sqrt(1 - alphas_bar)
+
+        # for reverse process p_theta(x_t-1 | x_t)
```
Comment (Ulvi-Movs) [inline diffusion.py:42]: same for here? i see x_t - 1 in formula, but i think i see in implementation 1 - x_t . Please correct me if I'm wrong.

File: diffusion.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+            return new_x
+        else:
+            return u_t
+
+
+    def get_loss(self, x_0: torch.Tensor) -> torch.Tensor:
+        """
+        Draw random timesteps for batch of images, perform forward process, predict noise, get MSE
+        """
+        batch_size = x_0.shape[0]
```
Comment (Ulvi-Movs) [inline diffusion.py:129]: why we get batch size from x_0 shape?

File: diffusion.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+        t = torch.randint(low=0, high=self.T, size=(batch_size,))
+
+        x_noisy, noise = self.sample_forward_process(x_0, t)
+        noise_pred = self.model(x_noisy)  # model(x_noisy, t)
+        loss = F.mse_loss(noise, noise_pred)
+
+        return loss
+
+
+    def train(self, dataloader: DataLoader, lr: float, epochs: int) -> None:
```
Comment (Ulvi-Movs) [inline diffusion.py:139]: i think we need add early_stoping flag in for True value implement early_soping

File: diffusion.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+        "Train Unet model"
+
+        optimizer = optim.Adam(self.model.parameters(), lr=lr)
```
Comment (Ulvi-Movs) [inline diffusion.py:142]: i think we can make ls - increase or decrease dynamically in training process.

File: notebook.ipynb
Comment (Ulvi-Movs) [inline notebook.ipynb:1]: Its good idea beside of loss try to see other possible metrics too. We see that loss is decrease but didnt understand what it give to us, and dont now wich parametr we need to change.

**Review by Ulvi-Movs — CHANGES_REQUESTED** [review]

**Review by KajetanOzog — APPROVED** [review]
